test: add unit test for class ComplexArray, and add the notes in comp…

pxlxingliang opened the pull request:
…lexarray.h

range: module_base/complexarray.h, module_base/test/complexarray_test.cpp

pxlxingliang committed:
test: add unit test for class ComplexArray, and add the notes in complexarray.h
range: module_base/complexarray.h, module_base/test/complexarray_test.cpp

caic99 commented on source/module_base/complexarray.h:
```diff
@@ -59,8 +59,14 @@ class ComplexArray
 		(const int ind1=0, const int ind2=0, const int ind3=0, const int ind4=0)const;
 //  const std::complex < double> &operator()(int, int, int, int, int)const;
 
+	//set all element to be {0.0,0.0}
```
typo: elements

caic99 commented on source/module_base/complexarray.h:
```diff
@@ -59,8 +59,14 @@ class ComplexArray
+	//set all element to be {0.0,0.0}
 	void zero_out(void);
+
+	//set all element multiply -1
```
typo: elements, same below
Maybe another expression, like "Multiply all elements by -1"? "Set" here may cause ambiguity.

caic99 commented on source/module_base/complexarray.h:
```diff
@@ -83,23 +89,37 @@ class ComplexArray
+void scaled_sum(std::complex < double> c1, const ComplexArray &cd1,
+           std::complex < double> c2, const ComplexArray &cd2,
+           ComplexArray &cd3);
+//void scaled_sum(double r1, ComplexArray &cd1,
```
Why commented codes here?

pxlxingliang replied:
the commented function has been declared at line 110. This should be a mistake of previous author, he may like to declare the function at line 115 which is actually defined in .cpp file.

caic99 replied:
> the commented function has been declared at line 110. This should be a mistake of previous author, he may like to declare the function at line 115 which is actually defined in .cpp file.

OK. If duplicated, just remove them.

caic99 commented on source/module_base/complexarray.h:
```diff
@@ -83,23 +89,37 @@ class ComplexArray
-                double r2, ComplexArray &cd2,
-                ComplexArray &cd3);
+/* Does cd3 = c1*cd1 + c2*cd2 */
+void scaled_sum(std::complex < double> c1, const ComplexArray &cd1,
```
I suggest no spaces between '<' and 'double'.

pxlxingliang replied:
agree.  I will fix it and also elsewhere.

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -2,14 +2,19 @@
 
 /**
-* - Tested Function:
+* - Tested Function of class ComplexArray:
```
maybe "functions"?

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -2,14 +2,19 @@
 /**
-* - Tested Function:
+* - Tested Function of class ComplexArray:
+*   - construction:
```
"Constructor"?

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -84,14 +107,78 @@ class ComplexArray_test : public testing::Test
 
-TEST_F(ComplexArray_test,operator_equal)
+
+TEST(ComplexArray, construction1_test)
```
Naming tests by a more meaningful string, rather than a number.

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -167,3 +290,148 @@ TEST_F(ComplexArray_test,operator_parentheses)
+    for (int i=0;i<a.getSize();++i)
+    {
+        EXPECT_TRUE((a.ptr[i].real() >= -0.5) && (a.ptr[i].imag() >= -0.5));
+        EXPECT_TRUE((a.ptr[i].real() < 0.5) && (a.ptr[i].imag() < 0.5));
```
What's the difference between line 325 and 326?

pxlxingliang replied:
Isn't that obvious?

caic99 replied:
> Isn't that obvious?

My bad, I didn't see the minus.

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -167,3 +290,148 @@ TEST_F(ComplexArray_test,operator_parentheses)
+
+}
+
+TEST(ComplexArray,fuction_getBoundSize)
```
I suggest declaiming "function" in test name is not needed. Just the function name itself is enough (and maybe function arguments, if any).

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -84,14 +107,78 @@ class ComplexArray_test : public testing::Test
+    ModuleBase::ComplexArray b = a2;
+    for ( int i = 0;i<a2.getSize();++i)
+    {
+        EXPECT_EQ(b.ptr[i],a2.ptr[i]);
```
Why not using `EXPECT_DOUBLE_EQ` if contents in `ComplexArray` class are stored in double?

pxlxingliang replied:
It is a complex value, and can not used in EXPECT_DOUBLE_EQ

complexarray_test.cpp(456): error: no instance of function template "testing::internal::CmpHelperFloatingPointEQ" matches the argument list
            argument types are: (const char [10], const char [10], std::complex<double>, std::complex<double>)
          EXPECT_DOUBLE_EQ(c2.ptr[i],com1*com2);

caic99 replied:
> It is a complex value, and can not used in EXPECT_DOUBLE_EQ
> 
> complexarray_test.cpp(456): error: no instance of function template "testing::internal::CmpHelperFloatingPointEQ" matches the argument list argument types are: (const char [10], const char [10], std::complex, std::complex) EXPECT_DOUBLE_EQ(c2.ptr[i],com1*com2);

I see. So, we have to compare the real part and imag part separately with `EXPECT_DOUBLE_EQ`?

caic99 commented on source/module_base/complexarray.h:
```diff
@@ -83,23 +89,37 @@ class ComplexArray
 add_scale_abs2(const std::complex < double> &c, const ComplexArray & in,
                ComplexArray &out);
 
+// Take "dot-product" of two ComplexArray:  sum of cd1(conjugate)[i] * cd2[i]
```
I'm sorry but why conjugate on cd1 in dot product?

pxlxingliang replied:
i also don't know why. this is just what the code represented.

caic99 replied:
> i also don't know why. this is just what the code represented.

That's weird, did this function used in code somewhere else?

pxlxingliang replied:
I can understand this action, which is really common in quantum chemistry like (psi^* * psi), of cause this is not a standard dot-product, so the comment is "dot-product". I do not know how to name this action, so I just copy it from .cpp file, and add my my understanding in the last half sentence .

caic99 replied:
> I can understand this action, which is really common in quantum chemistry like (psi^* * psi), of cause this is not a standard dot-product, so the comment is "dot-product". I do not know how to name this action, so I just copy it from .cpp file, and add my my understanding in the last half sentence .

OK, do we need a new name here, or leave it as is? I suggest changing a more unique name avoiding confusing.

pxlxingliang replied:
My previous understanding is wrong, after checking with wenfei-li,  this action is the standard dot-product of two complex array.

caic99 replied:
> My previous understanding is wrong, after checking with wenfei-li, this action is the standard dot-product of two complex array.

You are right.
Ref: https://www.mathworks.com/help/matlab/ref/dot.html
> When inputs A and B are real or complex vectors, the dot function treats them as column vectors and dot(A,B) is the same as sum(conj(A).*B).

caic99 commented on source/module_base/complexarray.h:
```diff
@@ -83,23 +89,37 @@ class ComplexArray
 void scale_accumulate(double r, const ComplexArray &cd1, ComplexArray &cd2);
+
+/* Does cd2 += c * cd1 */
 void scale_accumulate(std::complex < double> c, const ComplexArray &cd1, ComplexArray &cd2);
```
If possible, add `const` for those parameters not modified.

pxlxingliang replied:
it will modify cd2

caic99 replied:
> it will modify cd2

Actually `c` I mean

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -84,14 +107,78 @@ class ComplexArray_test : public testing::Test
-    EXPECT_EQ(b2,a2);
-    EXPECT_NE(c2,a4);
-    EXPECT_DEATH(c2=a4,"");
+    for(int i = 0;i<a2.getSize();++i)
```
Maybe add some comments here to notify that `a2=com1` was performed when constructing test class. Or just do it again.

pxlxingliang committed:
test: fix mistakes in complexarray_test.py and complexarray.h
	. fix some typos in complexarray.h
	. delete the space between "<" and "double" in "std::complex < double>"
	. delete the wrong declare of function scaled_sum
	. delete the assignment in ComplexArray_test::SetUp(), but do assignment in each TEST.
	. modify the test name about construction to be meaningful
range: module_base/complexarray.h module_base/test/complexarray_test.cpp

pxlxingliang committed:
test: fix com1*-1

pxlxingliang committed:
test: modify the comparsion of two complex
    by using EXPECT_NEAR to compare the real and imag partes of
    two complex, the thread is 1e-14.
range: module_base/test/complexarrray.cpp

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -2,14 +2,19 @@
 
 /************************************************
-*  unit test of class ComplexArray
+*  unit test of class ComplexArray and related fucntions
```
Typo: functions

pxlxingliang committed:
test: fix the typo
range: module_base/test/complexarray_test.cpp

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -45,125 +50,417 @@
+//compare two complex by using EXPECT_NEAR()
+void EXPECT_COMPLEXD(const std::complex<double>& a,const std::complex<double>& b)
+{
+    EXPECT_NEAR(a.real(),b.real(),prec_test);
```
Shall we use `EXPECT_DOUBLE_EQ` when comparing? What's the point of designating a precision threshold?

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -45,125 +50,417 @@
 
+double prec_test = 1e-14;
+//compare two complex by using EXPECT_NEAR()
+void EXPECT_COMPLEXD(const std::complex<double>& a,const std::complex<double>& b)
```
The name `EXPECT_COMPLEXD` should be more distinct. Maybe something like `EXPECT_COMPLEX_DOUBLE_NEAR` or `EXPECT_COMPLEX_EQUAL`, indicating type ("complex", "complex \<double>") and method("NEAR", "EQ").

pxlxingliang committed:
test: fix typoes, and modify the test name
range: module_base/test/complexarray_test.cpp

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -45,125 +50,419 @@
+{   
+    a2 = com1;
+    ModuleBase::ComplexArray b = a2;
+    for ( int i = 0;i<a2.getSize();++i)
```
Would you remove the space ahead `int`?

pxlxingliang replied:
ok

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -45,125 +50,419 @@
 */
 
 
+double prec_test = 1e-14;
```
You may remove redundant codes and variables: line 74, 78-79.

pxlxingliang replied:
ok

pxlxingliang committed:
test: modify the layout of some codes

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -45,125 +50,416 @@
+{
+    a2 = com1;
+    b2 = com2;
+    point_mult(a2,b2,c2);
```
The name `point_mult` is really weird here - being something sounds like "dot product" but actually no. Any suggestions?
> .* is Element-wise multiplication follow rules for array operations
 Also called: Hadamard Product, Schur Product and broadcast

pxlxingliang replied:
Agree, I think hadamard_product is good.

pxlxingliang replied:
I just searched it and found that there has another name that is pointwise multiplication. In that light, abbreviated as point_mult is not unacceptable.
https://math.stackexchange.com/questions/20412/element-wise-or-pointwise-operations-notation

caic99 commented on source/module_base/test/complexarray_test.cpp:
```diff
@@ -45,125 +50,416 @@
+
+}
+
+TEST(ComplexArray, constructor_rvalue_test)
```
We're already in test, do we need to emphasize in `constructor_rvalue_test`?

pxlxingliang replied:
I think this is not critical, and is just a habit, and will not cause any ambiguity or misunderstanding.
